File: app/scraper/info/igdb.py
import json
import logging
from datetime import datetime, timezone

import requests
from igdb.wrapper import IGDBWrapper
from unidecode import unidecode

from app.configparser import Config
from app.scraper.info.utils import RESULT_MATCH_THRESHOLD, get_match_score
from app.sqlalchemy import IgdbInfo

logger = logging.getLogger(__name__)


async def get_igdb_id(search_string: str) -> int | None:
    igdb = await get_igdb_wrapper()

    if igdb is None:
        return None

    logger.info(f"Getting id for {search_string}")

    # Replace non-Latin characters with their closest representation
    # and replace " because that would break the query
    api_search_string = unidecode(search_string).replace('"', "")
    try:
        raw_response: bytes = igdb.api_request(
            "games",
            f'search "{api_search_string}"; fields name; where version_parent = null; limit 50;',
        )

    except requests.exceptions.RequestException as e:
        logger.error(f"IGDB: Request failed: {e}")
        return None

    response = json.loads(raw_response)

    # Read all results and use the one with the highest difflib score (lower cased!)
    best_id: int | None = None
    best_score: float = 0
    best_title: str | None = None

    for entry in response:
        result_name = entry["name"]
        score = get_match_score(search_string, result_name)

        if score >= RESULT_MATCH_THRESHOLD and score > best_score:
            best_id = entry["id"]
            best_score = score
            best_title = result_name
            logger.debug(
                f"IGDB: Found match {result_name} with a score of {(score*100):.0f} %"
            )
        else:
            logger.debug(
                f"IGDB: Ignoring {result_name} as it's score of {(score*100):.0f} % is too low"
            )

    if best_id is not None:
        logger.info(
            f"IGDB: Search for {api_search_string} resulted in {best_title} ({best_id}) as the best match with a score of {(best_score*100):.0f} %"
        )
        return best_id

    logger.info(f"IGDB: Search for {api_search_string} found no result")

    return None


async def get_igdb_details(
    id_: int | None = None, title: str | None = None
) -> IgdbInfo | None:
    igdb_game_id: int | None = None

    if id_:
        igdb_game_id = id_

    if not igdb_game_id and title:
        igdb_game_id = await get_igdb_id(title)

    if not igdb_game_id:
        # No entry found, not adding any data
        return None

    logger.info(f"IGDB: Reading details for IGDB id {id_}")

    igdb_info = IgdbInfo()
    igdb_info.id = igdb_game_id

    igdb = await get_igdb_wrapper()

    if igdb is None:
        return None

    raw_response: bytes = igdb.api_request(
        "games",
        f"fields *; where id = {igdb_game_id};",
    )

    response = json.loads(raw_response)

    if len(response) == 0:
        return None

    try:
        igdb_info.name = response[0]["name"]
    except KeyError:
        pass

    try:
        igdb_info.url = response[0]["url"]
    except KeyError:
        pass

    try:
        igdb_info.short_description = response[0]["summary"]
    except KeyError:
        pass

    try:
        unix_releasedate = response[0]["first_release_date"]
        timestamp = datetime.utcfromtimestamp(unix_releasedate)
        igdb_info.release_date = timestamp.replace(tzinfo=timezone.utc)
    except KeyError:
        pass

    try:
        igdb_info.user_score = int(response[0]["rating"])
    except (KeyError, ValueError):
        pass

    try:
        igdb_info.user_ratings = response[0]["rating_count"]
    except KeyError:
        pass

    try:
        igdb_info.meta_score = int(response[0]["aggregated_rating"])
    except (KeyError, ValueError):
        pass

    try:
        igdb_info.meta_ratings = response[0]["aggregated_rating_count"]
    except KeyError:
        pass

    # TODO: Publisher, Genres, Cover
    # genres = List of genres (ids only, have to be called separately)
    # cover = Cover image of the game (id only)

    return igdb_info


# TODO: Only use one connection, rate limit to < 4 per second
async def get_igdb_wrapper() -> IGDBWrapper | None:
    client_id = Config.get().igdb_client_id
    client_secret = Config.get().igdb_client_secret

    # TODO: Make this asynchronously and add timeout
    r = requests.post(
        f"https://id.twitch.tv/oauth2/token?client_id={client_id}&client_secret={client_secret}&grant_type=client_credentials"
    )
    if r._content is None:
        logger.error("Failed to get IGDB wrapper.")
        return None
    access_token = json.loads(r._content)["access_token"]

    wrapper = IGDBWrapper(client_id, access_token)

    if wrapper is None:
        raise Exception("Could not get IGDB wrapper")

    return wrapper

File: src/services/gameinfo/igdb/igdb.ts
import type { NewIgdbInfo } from "@/types/database";
import { getMatchScore, normalizeString } from "@/utils";
import { logger } from "@/utils/logger";
import { DateTime } from "luxon";

interface IgdbAuth {
  accessToken: string;
  expiresAt: number;
}

interface IgdbGameResult {
  id: number; // IGDB ID
  name: string; // Game title
  url: string; // URL to the game on IGDB
  summary?: string; // Short description
  first_release_date?: number; // Release data as Unix timestamp (s)
  rating?: number; // User score
  rating_count?: number; // User ratings
  aggregated_rating?: number; // Meta score
  aggregated_rating_count?: number; // Meta ratings
}

interface IgdbSearchResult {
  id: number;
  name: string;
}

export class IgdbClient {
  private static readonly API_URL = "https://api.igdb.com/v4";
  private static readonly AUTH_URL = "https://id.twitch.tv/oauth2/token";
  private static readonly MINIMUM_ACCEPTABLE_SCORE = 0.75;

  private auth: IgdbAuth | null = null;

  constructor(
    private readonly clientId: string,
    private readonly clientSecret: string,
  ) {}

  /**
   * Search IGDB via the APIv4 and return the best match in the results.
   *
   * The comparison is done with difflib, lower cased.
   * @param searchString
   * @returns
   */
  public async searchGame(searchString: string): Promise<number | null> {
    await this.ensureAuth();

    const query = `search "${normalizeString(searchString)}";
fields name;
where version_parent = null;
limit 50;
`;

    logger.debug(`Searching for game: ${searchString} with query: ${query}`);

    const results = await this.apiRequest<IgdbSearchResult[]>("games", query);

    if (!results.length) return null;

    let bestMatch: { id: number; score: number; title: string } | null = null;
    for (const game of results) {
      const score = getMatchScore(searchString, game.name);
      if (
        score >= IgdbClient.MINIMUM_ACCEPTABLE_SCORE &&
        (!bestMatch || score > bestMatch.score)
      ) {
        bestMatch = { id: game.id, score, title: game.name };
        logger.debug(
          `Found match ${game.name} with score ${(score * 100).toFixed()}%`,
        );
      } else {
        logger.debug(
          `Rejected match ${game.name} as it's score of ${(score * 100).toFixed()}% is too low.`,
        );
      }
    }

    if (!bestMatch) {
      logger.debug(`No match found for ${searchString}.`);
      return null;
    }

    logger.debug(
      `Best match for ${searchString} is ${bestMatch.title} with score ${(bestMatch.score * 100).toFixed()}%.`,
    );
    return bestMatch.id;
  }

  public async getDetails(gameId: number): Promise<NewIgdbInfo | null> {
    await this.ensureAuth();

    const query = `
      fields *;
      where id = ${gameId.toFixed()};
    `;

    const results = await this.apiRequest<IgdbGameResult[]>("games", query);

    if (!results.length) return null;

    // We're searching by ID, so we only get one result
    const [game] = results;

    return {
      id: game.id,
      name: game.name,
      url: game.url,
      short_description: game.summary ?? null,
      release_date: game.first_release_date
        ? DateTime.fromSeconds(game.first_release_date).toISO()
        : null,
      user_score: game.rating ? Math.round(game.rating) : null,
      user_ratings: game.rating_count ?? null,
      meta_score: game.aggregated_rating
        ? Math.round(game.aggregated_rating)
        : null,
      meta_ratings: game.aggregated_rating_count ?? null,
    };
  }

  private async ensureAuth(): Promise<void> {
    if (this.auth) {
      const expiresSoon =
        DateTime.now() <
        DateTime.fromMillis(this.auth.expiresAt).minus({ minutes: 1 });
      if (expiresSoon) return;
    }

    const params = new URLSearchParams({
      client_id: this.clientId,
      client_secret: this.clientSecret,
      grant_type: "client_credentials",
    });

    const response = await fetch(IgdbClient.AUTH_URL, {
      method: "POST",
      body: params,
    });

    if (!response.ok) {
      throw new Error(`IGDB auth failed: ${response.statusText}`);
    }

    const { access_token, expires_in } = (await response.json()) as {
      access_token: string;
      expires_in: number;
    };

    this.auth = {
      accessToken: access_token,
      expiresAt: DateTime.now().plus({ seconds: expires_in }).toMillis(),
    };
  }

  private async apiRequest<T>(endpoint: string, query: string): Promise<T> {
    if (!this.auth) throw new Error("Not authenticated");

    const response = await fetch(`${IgdbClient.API_URL}/${endpoint}`, {
      method: "POST",
      headers: {
        "Client-ID": this.clientId,
        Authorization: `Bearer ${this.auth.accessToken}`,
      },
      body: query,
    });

    logger.debug(
      `IGDB API request to ${endpoint} returned status ${response.status.toFixed()} (ok: ${response.ok ? "yes" : "no"}).`,
    );
    if (!response.ok) {
      throw new Error(`IGDB API error: ${response.statusText}`);
    }

    return response.json() as Promise<T>;
  }
}
